README.MD

```diff
@@ -213,13 +213,20 @@ duplicate HRID, the module will reject the submission.
 
 # Domain event pattern
 
-The pattern means that every time when an instance is created/updated/removed
-a message is posted to `inventory.instance` kafka topic with following structure:
+The pattern means that every time when an instance/item is created/updated/removed
+a message is posted to kafka topic:
+* `inventory.instance` - for instances;
+* `inventory.item` - for items;
+
+The event payload has following structure:
 ```javascript
 {
-  "old": {...}, // the instance before update or delete
-  "new": {...}, // the instance after update or create
+  "old": {...}, // the instance/item before update or delete
+  "new": {...}, // the instance/item after update or create
   "type": "UPDATE|DELETE|CREATE", // type of the event
   "tenant": "diku" // tenant name
 }
 ```
+
+Kafka partition key for the both topics is instance id (for items it is retrieved from
+associated holding record).
```

Comment:
Which instance ID is used as the partition key when the change to an item moves the item between two instances (via a move between holdings)? Is it the instance ID before the change or after?

Reply:
It is instance id after the change.

Reply:
If the elastic search import focuses on the instance ID, how will the item be removed from the projection of the information for the instance before the change?

Reply:
It is probably responsibility of the new `mod-search` service, it should figure it out. However, we haven't fully agreed final index structure for the instances and items yet, there was an idea to have a separate index for instance, and in this case we don't need this manipulation at all.

Reply:
> It is probably responsibility of the new `mod-search` service, it should figure it out. 

Agreed. I only raise it here because the designs of both likely need to be complementary. Decisions here could introduce constraints for the way mod-search is implemented.

> However, we haven't fully agreed final index structure for the instances and items yet, there was an idea to have a separate index for instance, and in this case we don't need this manipulation at all.

In which case, why partition items by instance ID?

I don't think I follow this.

As I understand it, from my basic knowledge of Elastic Search, searches to be performant they typically need to be fulfilled by a single index. And searches in inventory, like find the copy of Harry Potter and the Goblet of Fire with barcode 12345 rely on a combination of instance and item information.

I would imagine these kinds of searches wouldn't be well fulfilled if that information is split across indexes.

Reply:
Yeah, you're right. But Mikhail mentioned that if we use a single index for instance with item nested, than a search query, for example: `instance.title=="..." and item.status=="..."` most likely will work as an OR (will return instances even though they don't have items with the status). But maybe I'm missing something or misunderstood it. That's why I said that we haven't agreed final structure of the index.

> In which case, why partition items by instance ID?

In any case we must keep this relation in order to allow this item and instance searches.

Reply:
> But Mikhail mentioned that if we use a single index for instance with item nested, than a search query, for example: `instance.title=="..." and item.status=="..."` most likely will work as an OR (will return instances even though they don't have items with the status). But maybe I'm missing something or misunderstood it. That's why I said that we haven't agreed final structure of the index.

Hmm, ok. Based upon my understanding of what the users want, that would be incorrect. I may well have not understood.

> > In which case, why partition items by instance ID?
> 
> In any case we must keep this relation in order to allow this item and instance searches.

Which brings me back to the challenge of making sure that the instance the item used to belong to doesn't erroneously appear in searches involving items.

Reply:
@bohdan-suprun @MikhailFokanov 

> Yeah, you're right. But Mikhail mentioned that if we use a single index for instance with item nested, than a search query, for example: `instance.title=="..." and item.status=="..."` most likely will work as an OR (will return instances even though they don't have items with the status). But maybe I'm missing something or misunderstood it. That's why I said that we haven't agreed final structure of the index.

According to [this comment](https://wiki.folio.org/pages/viewpage.action?pageId=52135163&focusedCommentId=54886420#comment-54886420) by @MikhailFokanov there is intended to be an index that has both instance and item properties.

How will the index entry for the prior instance be updated with these messages?

Reply:
We haven't started implementation of items, so I can not answer your question. But there definitely should be some linkage between instances and items, we will make sure that item is detached from the old instance and holding.

Reply:
Had a chat with Mikhail, we will also have to send instanceId in the new/old item. I will create a separate story for this - https://issues.folio.org/browse/MODINVSTOR-649